pkg/binding/message.go

```diff
@@ -0,0 +1,96 @@
+package binding
+
+import "context"
+
+// Message is the interface to a binding-specific message containing an event.
+//
+// Reliable Delivery
+//
+// There are 3 reliable qualities of service for messages:
+//
+// 0/at-most-once/unreliable: messages can be dropped silently.
+//
+// 1/at-least-once: messages are not dropped without signaling an error
+// to the sender, but they may be duplicated in the event of a re-send.
+//
+// 2/exactly-once: messages are never dropped (without error) or
+// duplicated, as long as both sending and receiving ends maintain
+// some binding-specific delivery state. Whether this is persisted
+// depends on the configuration of the binding implementations.
+//
+// The Message interface supports QoS 0 and 1, the ExactlyOnceMessage interface
+// supports QoS 2
+//
+// The Structured and Binary methods allows to perform an optimized encoding of a Message
+// to a specific data structure. A Sender should try each method of interest and fall back to ToEvent() if none are supported.
+// For encoding a message, look at binding.Encode function.
+//
+// Every binding.Message implementation must specify if the message can be accessed one or more times.
+//
+// When a Message can be forgotten by the entity who produced the message, Message.Finish() *must* be invoked.
+type Message interface {
+	// Return the type of the message Encoding.
+	// The encoding should be preferably computed when the message is constructed.
+	Encoding() Encoding
+
+	// Structured transfers a structured-mode event to a StructuredEncoder.
+	// It must return ErrNotStructured if message is not in structured mode.
+	//
+	// Returns a different err if something wrong happened while trying to read the structured event.
+	// In this case, the caller must Finish the message with appropriate error.
+	//
+	// This allows Senders to avoid re-encoding messages that are
+	// already in suitable structured form.
+	Structured(context.Context, StructuredEncoder) error
```

Comment:
I have tried but I have no idea how to use this method or the binary method because I do not know how to make a StructuredEncoder.

Reply:
Both `StructuredEncoder` and `BinaryEncoder` are provided by the specific binding implementation. Their functionality is usually "served" through an `Encode` method, like:

* https://github.com/cloudevents/sdk-go/blob/master/pkg/bindings/kafka_sarama/producer_message_encoder.go#L23
* https://github.com/cloudevents/sdk-go/blob/master/pkg/bindings/amqp/encoder.go#L18
* https://github.com/cloudevents/sdk-go/blob/master/pkg/bindings/http/response_encoder.go#L17
* https://github.com/cloudevents/sdk-go/blob/master/pkg/bindings/http/request_encoder.go#L17

So the bindings user don't need to invoke them directly